refactor: use ROWS and COLS in Position.ts

wuzengfu opened the pull request.

kspc1000 committed:
feat(logic, piece): Added Advisor.ts

kspc1000 committed:
📝docs(logic): Add comments to Pawn.ts .filter methods

kspc1000 committed:
✅test(logic, piece): Added Advisor.test.ts

Finished testing Advisor in an empty board

wuzengfu committed:
build: add set_up_husky script

wuzengfu committed:
ci: update github action script

wuzengfu committed:
refactor: use ROWS and COLS in Position.ts

wuzengfu committed:
refactor: use ROWS and COLS in Position.ts

kspc1000 commented on src/logic/Position.ts:
```diff
@@ -30,7 +32,7 @@ class Position {
     }
 
-    return !(this.x >= 9 || this.y >= 10)
+    return !(this.x >= Position.COLS || this.y >= Position.ROWS)
```
I was thinking maybe instead we can do this

```ts
const boundaryConditions: boolean[] = [
  this.x >= 0,
  this.y >= 0,
  this.x < Position.COLS,
  this.y < Position.ROWS
]
    
return boundaryConditions.every((condition) => condition)
```

kspc1000 replied:
Maybe also add a `Position.ZERO` or `Position.START` variable or something 

```ts
class Position {
  public static ZERO = 0
}
```

kspc1000 reviewed (commented):
test

kspc1000 commented on src/logic/Position.ts:
```diff
@@ -30,7 +32,7 @@ class Position {
     }
 
-    return !(this.x >= 9 || this.y >= 10)
+    return !(this.x >= Position.COLS || this.y >= Position.ROWS)
```
I was also thinking we could technically do something like this
```ts
return (
  this.x >= 0
  && this.y >= 0
  && this.x < Position.COLS
  && this.y < Position.ROWS
)
```
but our current **ESlint** configurations doesn't allow this

wuzengfu replied:
> I was also thinking we could technically do something like this
> 
> ```ts
> return (
>   this.x >= 0
>   && this.y >= 0
>   && this.x < Position.COLS
>   && this.y < Position.ROWS
> )
> ```
> 
> but our current **ESlint** configurations doesn't allow this

Usually we will try to split the conditions if exceed 3.

wuzengfu replied:
> Maybe also add a `Position.ZERO` or `Position.START` variable or something
> 
> ```ts
> class Position {
>   public static ZERO = 0
> }
> ```

yeah, this is feasible, we can introduce a `START`

wuzengfu replied:
> I was thinking maybe instead we can do this instead. It's easier to read each condition line by line so I thought of this:
> 
> ```ts
> const boundaryConditions: boolean[] = [
>   this.x >= 0,
>   this.y >= 0,
>   this.x < Position.COLS,
>   this.y < Position.ROWS
> ]
>     
> return boundaryConditions.every((condition) => condition)
> ```

I usually write this only when it is hard to partition conditions, but for this case, we can easily partition it into two major sub-conditions:
1. Check left boundary (x < 0).
2. Check right boundary (x >= COLS).

Also note that this trick only works for boolean expression in [CNF](https://www.geeksforgeeks.org/normal-and-principle-forms#conjunctive-normal-form-cnf) form

kspc1000 replied:
I was thinking `Position.ZERO` would be a better name because technically the pieces don't actually "start" at 0 but it just happens to be the bottom left corner of the chessboard.

kspc1000 replied:
[Repost] I was thinking maybe instead we can do this instead. It's easier to read each condition line by line so I thought of this:

```ts
const boundaryConditions: boolean[] = [
  this.x >= 0,
  this.y >= 0,
  this.x < Position.COLS,
  this.y < Position.ROWS
]
    
return boundaryConditions.every((condition) => condition)
```

_Reposting this comment because I accidentally deleted it._

kspc1000 replied:
> > I was thinking maybe instead we can do this instead. It's easier to read each condition line by line so I thought of this:
> > 
> > ```ts
> > const boundaryConditions: boolean[] = [
> >   this.x >= 0,
> >   this.y >= 0,
> >   this.x < Position.COLS,
> >   this.y < Position.ROWS
> > ]
> >     
> > return boundaryConditions.every((condition) => condition)
> > ```
> 
> I usually write this only when it is hard to partition conditions, but for this case, we can easily partition it into two major sub-conditions:
> 1. Check left boundary (x < 0).
> 2. Check right boundary (x >= COLS).
> 
> Also note that this trick only works for boolean expression in [CNF](https://www.geeksforgeeks.org/normal-and-principle-forms#conjunctive-normal-form-cnf) form

Yeah test

kspc1000 replied:
> > I was thinking maybe instead we can do this instead. It's easier to read each condition line by line so I thought of this:
> > 
> > ```ts
> > const boundaryConditions: boolean[] = [
> >   this.x >= 0,
> >   this.y >= 0,
> >   this.x < Position.COLS,
> >   this.y < Position.ROWS
> > ]
> >     
> > return boundaryConditions.every((condition) => condition)
> > ```
> 
> I usually write this only when it is hard to partition conditions, but for this case, we can easily partition it into two major sub-conditions:
> 1. Check left boundary (x < 0).
> 2. Check right boundary (x >= COLS).
> 
> Also note that this trick only works for boolean expression in [CNF](https://www.geeksforgeeks.org/normal-and-principle-forms#conjunctive-normal-form-cnf) form

Yeah I was thinking about this last night as well, it's very clean and explicit:

```ts
const top = this.y < Position.ROWS
const right = this.x < Position.COLS
const bottom = this.y >= Position.ZERO
const left = this.x >= Position.ZERO
return [top, right, bottom, left].every(Boolean)
```

wuzengfu replied:
> I was thinking `Position.ZERO` would be a better name because technically the pieces don't actually "start" at 0 but it just happens to be the bottom left corner of the chessboard.

This is `Position` class, it has nothing to do with the chessboard. The word "START", in this case, means the x-axis and y-axis starts at `START`. Also, we want to extract "0" to a constant variable because we want to make the meaning of "0" clear so that everyone knows what the number "0" means. If we used `Position.ZERO`, then it's still "0", it would make this constant extraction pointless since everyone knows `ZERO` means `0`, then why not just `0`? In this case, `0` is a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)).

wuzengfu replied:
> > > I was thinking maybe instead we can do this instead. It's easier to read each condition line by line so I thought of this:
> > > ```ts
> > > const boundaryConditions: boolean[] = [
> > >   this.x >= 0,
> > >   this.y >= 0,
> > >   this.x < Position.COLS,
> > >   this.y < Position.ROWS
> > > ]
> > >     
> > > return boundaryConditions.every((condition) => condition)
> > > ```
> > 
> > 
> > I usually write this only when it is hard to partition conditions, but for this case, we can easily partition it into two major sub-conditions:
> > 
> > 1. Check left boundary (x < 0).
> > 2. Check right boundary (x >= COLS).
> > 
> > Also note that this trick only works for boolean expression in [CNF](https://www.geeksforgeeks.org/normal-and-principle-forms#conjunctive-normal-form-cnf) form
> 
> Yeah I was thinking about this last night as well, it's very clean and explicit:
> 
> ```ts
> const top = this.y < Position.ROWS
> const right = this.x < Position.COLS
> const bottom = this.y >= Position.ZERO
> const left = this.x >= Position.ZERO
> return [top, right, bottom, left].every(Boolean)
> ```

I know this trick looks impressive for most of developers, but in reality, we only use this if we really cannot partition our conditions nicely, since there are just four conditions and can be esaily divided into two, it's better to use the traditional way.

wuzengfu replied:
> > I was thinking `Position.ZERO` would be a better name because technically the pieces don't actually "start" at 0 but it just happens to be the bottom left corner of the chessboard.
> 
> This is `Position` class, it has nothing to do with the chessboard. The word "START", in this case, means the x-axis and y-axis starts at `START`. Also, we want to extract "0" to a constant variable because we want to make the meaning of "0" clear so that everyone knows what the number "0" means. `0` in this case is a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)). If we used `Position.ZERO`, then it's still "0", it would make this constant extraction pointless since everyone knows `ZERO` means `0`, then why not just `0`?

kspc1000 committed:
✅ test(logic, piece): Added "Advisor surrounded by obstacles" for Advisor.test.ts

kspc1000 committed:
✨feat(logic, piece): Added AdvisorFactory.ts

kspc1000 committed:
✅ test(logic, piece): Added AdvisorFactory.test.ts

kspc1000 committed:
🐛fix(test, piece): Wrote wrong numbers for a few test cases

kspc1000 committed:
✅test(logic, piece): Added test case for red advisor at (4, 1) for Advisor.test.ts

This test ensures that the advisor can move towards 4 directions when in the middle of an EMPTY palace

kspc1000 committed:
✏️fix(piece): Typo in AdvisorFactory comment

kspc1000 committed:
♻️refactor(logic, piece): Removed redundant comments in Advisor.ts

kspc1000 committed:
🐛 fix(formatting): Added endOfLine: auto

kspc1000 committed:
Merge pull request #20 from xiangqi-master/bugfix/prettier-delete-line-break-warnings

🐛 fix: Resolve unnecessary 'Delete CR' warnings across the codebase

wuzengfu commented:
Add a `Position.START=0`

kspc1000 committed:
Merge branch 'develop' into 5-logic-for-advisor

kspc1000 committed:
fix: forgotten Advisor export

kspc1000 committed:
refactor(logic): Added START variable in Position.ts

kspc1000 committed:
refactor(logic): Added START variable in Position.ts

kspc1000 committed:
fix(test): Remove old and redundant "Test: isCrossRiver" in Position.test.ts

kspc1000 committed:
fix(test): Remove old and redundant "Test: isCrossRiver" in Position.test.ts

kspc1000 committed:
refactor(logic): Removed the need to apply De Morgan's Law in Position.ts

kspc1000 committed:
refactor(logic): Removed the need to apply De Morgan's Law in Position.ts

kspc1000 committed:
Merge pull request #16 from xiangqi-master/5-logic-for-advisor

kspc1000 committed:
Merge pull request #17 from xiangqi-master/refine-husky

wuzengfu committed:
docs: add github issue templates

wuzengfu committed:
Merge pull request #25 from xiangqi-master/add-issue-template

docs: add github issue templates

wuzengfu committed:
docs: update github issue templates for bug and feature

wuzengfu committed:
refactor: use Position.ROWS and Position.COLs in Rook.ts

wuzengfu committed:
Merge remote-tracking branch 'origin/refine-position-class' into refine-position-class
